processor/src/networks/bitcoin.rs

```diff
@@ -450,6 +477,85 @@ impl Bitcoin {
       }
     }
   }
+
+  // expected script has to start with SHA256 PUSH MSG_HASH OP_EQ ..
+  fn expected_script_pattern(&self, script: &ScriptBuf) -> Option<bool> {
+    let mut ins = script.instructions();
+
+    // first item should be SHA256 code
+    if ins.next()?.ok()?.opcode()? != OP_SHA256 {
+      return Some(false);
+    }
+
+    // next should be a data push
+    ins.next()?.ok()?.push_bytes()?;
+
+    // next should be a equality check
+    if ins.next()?.ok()?.opcode()? != OP_EQUAL {
+      return Some(false);
+    }
+
+    Some(true)
+  }
+
+  async fn spent_output_of(&self, input: &TxIn) -> TxOut {
+    let mut spent_tx = input.previous_output.txid.as_raw_hash().to_byte_array();
+    spent_tx.reverse();
+    let mut tx;
+    while {
+      tx = self.get_transaction(&spent_tx).await;
+      tx.is_err()
+    } {
+      log::error!("couldn't get transaction from bitcoin node: {tx:?}");
+      sleep(Duration::from_secs(5)).await;
+    }
+    tx.unwrap().output.swap_remove(usize::try_from(input.previous_output.vout).unwrap())
+  }
+
+  async fn extract_serai_data(&self, tx: &Transaction) -> Vec<u8> {
+    // check outputs
+    let mut data = (|| {
+      for output in &tx.output {
+        if output.script_pubkey.is_op_return() {
+          match output.script_pubkey.instructions_minimal().last() {
+            Some(Ok(Instruction::PushBytes(data))) => return data.as_bytes().to_vec(),
+            _ => continue,
+          }
+        }
+      }
+      vec![]
+    })();
+
+    // check inputs
+    if data.is_empty() {
+      for input in &tx.input {
+        if self.spent_output_of(input).await.script_pubkey.is_p2wsh() {
+          let witness = input.witness.to_vec();
+          let redeem_script = ScriptBuf::from_bytes(witness[1].clone());
```

Comment:
[1] will cause a panic. This needs to be last() with a continue if it doesn't exist (somehow).